pkg/federate/kubefed/cluster.go

```diff
@@ -1,9 +1,314 @@
 package kubefed
 
 import (
-	"github.com/submariner-io/admiral/pkg/federate"
+	"context"
+	"fmt"
+	"reflect"
+	"time"
+
+	federate "github.com/submariner-io/admiral/pkg/federate"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/klog"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	fedv1 "sigs.k8s.io/kubefed/pkg/apis/core/v1beta1"
+	"sigs.k8s.io/kubefed/pkg/client/generic/scheme"
+)
+
+const (
+	// The size of the event queue buffer for each ClusterEventHandler. This is largely a heuristic - we want a
+	// value reasonably large enough to avoid blocking the sender on bursts (eg when notifying of initial clusters
+	// on ClusterEventHandler registration) but not too large to risk running out of memory due to a slow handler.
+	defaultEventQueueBufferSize = 5000
+	defaultEnqueueTimerInterval = time.Second * 10
 )
 
 func (f *federator) WatchClusters(handler federate.ClusterEventHandler) error {
-	panic("not implemented")
+	f.startInformerOnce.Do(func() {
+		go func() {
+			f.kubeFedClusterInformer.Run(f.stopChan)
+			klog.Info("KubeFedCluster informer stopped")
+		}()
+
+		klog.Info("Started KubeFedCluster informer")
+	})
+
+	f.Lock()
+	defer f.Unlock()
+
+	// Use the clusterMap size if > the defaultEventQueueBufferSize
+	eventQueueBufferSize := len(f.clusterMap)
+	if eventQueueBufferSize < defaultEventQueueBufferSize {
+		eventQueueBufferSize = defaultEventQueueBufferSize
+	}
+
+	clusterWatcher := &clusterWatcher{
+		eventQueue: make(chan func(), eventQueueBufferSize),
+		stopChan:   f.stopChan,
+		handler:    handler,
+	}
+
+	f.clusterWatchers = append(f.clusterWatchers, clusterWatcher)
+
+	go clusterWatcher.eventLoop()
+
+	for clusterID, kubeConfig := range f.clusterMap {
+		clusterWatcher.onAdd(clusterID, kubeConfig)
+	}
+
+	klog.Infof("Federator added ClusterEventHandler \"%s\"", toString(handler))
```

Comment:
Let go handle the "toString" in whatever representation it comes out with. If we then see it doesn't suits the needs then we change it.

Reply:
I changed the logging to use %T which just prints the type. I was trying to duplicate the behavior of Java 's toString where you get the custom behavior if overridden else the default behavior of FQ class name. But I think just the type name should suffice for identification in log output.